From b1406840d3f3aecc8a809cd63036d59f8e4d6913 Mon Sep 17 00:00:00 2001
From: Ikey Doherty <ikey@solus-project.com>
Date: Sun, 28 Jan 2018 19:31:35 +0000
Subject: [PATCH] snap-confine/nvidia: Support legacy biarch trees for GLVND
 systems

GLVND systems are well catered for on biarch configurations, however the
older 304 + 340 series will rely on ld.so.conf tricks to mask the system
libGL.so with one in the private `libdir/nvidia*/` files. This is seen
consistently across Arch, Fedora and Solus. As such we allow these fallback
paths to work so that users of snapd on glvnd enabled distros with the older
legacy drivers can still enjoy snapd + NVIDIA together.

Signed-off-by: Ikey Doherty <ikey@solus-project.com>
---
 cmd/snap-confine/mount-support-nvidia.c | 8 ++++++++
 interfaces/builtin/opengl.go            | 2 +-
 2 files changed, 9 insertions(+), 1 deletion(-)

diff --git a/cmd/snap-confine/mount-support-nvidia.c b/cmd/snap-confine/mount-support-nvidia.c
index 42f56c3dc..9664fb880 100644
--- a/cmd/snap-confine/mount-support-nvidia.c
+++ b/cmd/snap-confine/mount-support-nvidia.c
@@ -65,6 +65,10 @@ static const size_t vulkan_globs_len =
 // FIXME: this doesn't yet work with libGLX and libglvnd redirector
 // FIXME: this still doesn't work with the 361 driver
 static const char *nvidia_globs[] = {
+	"/usr/lib/nvidia*/libGL.so*"
+	"/usr/lib/nvidia*/libEGL.so*"
+	"/usr/lib/nvidia*/libGLESv1_CM.so*",
+	"/usr/lib/nvidia*/libGLESv2.so*",
 	"/usr/lib/libEGL.so*",
 	"/usr/lib/libEGL_nvidia.so*",
 	"/usr/lib/libGL.so*",
@@ -103,6 +107,10 @@ static const size_t nvidia_globs_len =
 
 // 32-bit variants of the NVIDIA driver libraries
 static const char *nvidia_globs32[] = {
+	"/usr/lib32/nvidia*/libGL.so*"
+	"/usr/lib32/nvidia*/libEGL.so*"
+	"/usr/lib32/nvidia*/libGLESv1_CM.so*",
+	"/usr/lib32/nvidia*/libGLESv2.so*",
 	"/usr/lib32/libEGL.so*",
 	"/usr/lib32/libEGL_nvidia.so*",
 	"/usr/lib32/libGL.so*",
diff --git a/interfaces/builtin/opengl.go b/interfaces/builtin/opengl.go
index bd6588157..362583527 100644
--- a/interfaces/builtin/opengl.go
+++ b/interfaces/builtin/opengl.go
@@ -48,7 +48,7 @@ const openglConnectedPlugAppArmor = `
 /var/lib/snapd/hostfs/usr/share/vulkan/icd.d/*nvidia*.json r,
 
 # Main bi-arch GL libraries
-/var/lib/snapd/hostfs/{,usr/}lib{,32,64,x32}/{,@{multiarch}/}lib{GL,EGL,GLX}.so{,.*} rm,
+/var/lib/snapd/hostfs/{,usr/}lib{,32,64,x32}/{,@{multiarch}/}{,nvidia*/}lib{GL,EGL,GLX}.so{,.*} rm,
 
 /dev/dri/ r,
 /dev/dri/card0 rw,
-- 
2.16.1

